Model/Payment.php

```diff
@@ -1,199 +1,192 @@
+                'callback_url' => $this->urlBuilder->getUrl(
+                    'coingate/payment/callback',
+                    [
+                        '_query' => [
+                            'token' => $payment->getAdditionalInformation(self::COINGATE_ORDER_TOKEN_KEY)
+                        ]
+                    ]
+                ),
+                'cancel_url' => $this->urlBuilder->getUrl('coingate/payment/cancelOrder'),
+                'success_url' => $this->urlBuilder->getUrl('checkout/onepage/success'),
+                'title' => $this->storeManager->getWebsite()->getName(),
+                'description' => implode(', ', $description),
+                'token' => $payment->getAdditionalInformation(self::COINGATE_ORDER_TOKEN_KEY)
             ];
+        } catch (LocalizedException $exception) {
+            $this->logger->critical($exception->getMessage());
+
+            return null;
+        }
+
+        $client = $this->getClient();
+
+        try {
+            $cgOrder = $client->order->create($params);
+        } catch (ApiErrorException $exception) {
+            $this->logger->critical($exception->getMessage());
+
+            return null;
         }
+
+        return $cgOrder;
     }
 
     /**
+     * Validate CoinGate Callback
+     *
      * @param Order $order
+     * @param int $requestId
+     *
+     * @return void
      */
-    public function validateCoinGateCallback(Order $order)
+    public function validateCoinGateCallback(Order $order, int $requestId): void
     {
-
         try {
-            if (!$order || !$order->getIncrementId()) {
-                $request_order_id = (filter_input(INPUT_POST, 'order_id')
-                    ? filter_input(INPUT_POST, 'order_id') : filter_input(INPUT_GET, 'order_id')
-                );
-
-                throw new \Exception('Order #' . $request_order_id . ' does not exists');
-            }
-
-            $payment = $order->getPayment();
-            $get_token = filter_input(INPUT_GET, 'token');
-            $token1 = $get_token ? $get_token : '';
-            $token2 = $payment->getAdditionalInformation('coingate_order_token');
-
-            if ($token2 == '' || $token1 != $token2) {
-                throw new \Exception('Tokens do match.');
-            }
-
-            $request_id = (filter_input(INPUT_POST, 'id')
-                ? filter_input(INPUT_POST, 'id') :  filter_input(INPUT_GET, 'id'));
-            $cgOrder = \CoinGate\Merchant\Order::find($request_id);
+            $client = $this->getClient();
+            $cgOrder = $client->order->get($requestId);
 
             if (!$cgOrder) {
-                throw new \Exception('CoinGate Order #' . $request_id . ' does not exist');
+                throw new Exception('CoinGate Order #' . $requestId . ' does not exist');
             }
 
-            if ($cgOrder->status == 'paid') {
+            if ($cgOrder->status === self::PAID_STATUS) {
```

Comment:
maybe a check `paymentMethod == 'coingate' `is needed ?

previously we had issues that the customer selected coingate as its first payment method, then canceled it , selected another payment method but out callback managed to cancel the order with a different payment method.